fish-rust/src/builtins/string.rs

```diff
@@ -287,82 +284,79 @@ fn escape_code_length(code: &wstr) -> Option<usize> {
 /// This is probably down to some pipe buffer or some such,
 /// but too small means we need to call `read(2)` and str2wcstring a lot.
 const STRING_CHUNK_SIZE: usize = 1024;
-fn arguments<'iter, 'args>(
-    args: &'iter [&'args wstr],
+fn arguments<'iter>(
+    args: &'iter [WString],
     argidx: &'iter mut usize,
-    streams: &mut io_streams_t,
-) -> Arguments<'args, 'iter> {
+    streams: &mut IoStreams<'_>,
+) -> Arguments<'iter> {
     Arguments::new(args, argidx, streams, STRING_CHUNK_SIZE)
 }
 
 /// The string builtin, for manipulating strings.
-pub fn string(
-    parser: &mut parser_t,
-    streams: &mut io_streams_t,
-    args: &mut [&wstr],
-) -> Option<c_int> {
-    let cmd = args[0];
+pub fn string(parser: &Parser, streams: &mut IoStreams<'_>, args: &mut [WString]) -> Option<c_int> {
     let argc = args.len();
 
     if argc <= 1 {
+        let cmd = &args[0];
         streams
             .err
-            .append(wgettext_fmt!(BUILTIN_ERR_MISSING_SUBCMD, cmd));
-        builtin_print_error_trailer(parser, streams, cmd);
+            .append(&wgettext_fmt!(BUILTIN_ERR_MISSING_SUBCMD, cmd));
+        builtin_print_error_trailer(parser, streams.err, cmd);
         return STATUS_INVALID_ARGS;
     }
 
     if args[1] == "-h" || args[1] == "--help" {
+        let cmd = &args[0];
         builtin_print_help(parser, streams, cmd);
         return STATUS_CMD_OK;
     }
 
-    let subcmd_name = args[1];
+    let (cmd_names, args) = args.split_at_mut(2);
+    let cmd = &cmd_names[0];
+    let subcmd_name = &cmd_names[1];
 
-    match subcmd_name.to_string().as_str() {
-        "collect" => collect::Collect::default().run(parser, streams, args),
-        "escape" => escape::Escape::default().run(parser, streams, args),
-        "join" => join::Join::default().run(parser, streams, args),
+    let mut subcmd: Box<dyn StringSubCommand<'_>> = match subcmd_name.to_string().as_str() {
```

Comment:
https://github.com/fish-shell/fish-shell/commit/d6b41b33f4c888bbce3cba1e14e3dbf9dadc0afd#r124996810

> I wrote: @krobelus I disagree with using dynamic dispatch here, that change is noise and I don't see any good reasoning for it. I do have a regret from the string-rewrite in that it mixes data-and-behaviour — each subcommand-module should rather only expose `BuildinCmd`s, not any structs, which removes all the repeated arguments without using dynamic dispatch.

> @krobelus responded: I want to only construct the wgetopter_t in one place, not once per subcommand. Maybe this needs a quick macro

Reply:
I don't understand what you mean by constructing `wgetopter_t` in only one place, IMO this change should just be reverted.